src/Microsoft.ML.Transforms/StatefulFilterTransform.cs

```diff
@@ -171,20 +172,14 @@ public Cursor(StatefulFilterTransform<TSrc, TDst, TState> parent, RowCursor<TSrc
                 CursorChannelAttribute.TrySetCursorChannel(_parent.Host, _dst, Ch);
                 CursorChannelAttribute.TrySetCursorChannel(_parent.Host, _state, Ch);
 
-                if (parent._initStateAction != null)
-                    parent._initStateAction(_state);
+                parent._initStateAction?.Invoke(_state);
 
                 var appendedDataView = new DataViewConstructionUtils.SingleRowLoopDataView<TDst>(parent.Host, _parent._addedSchema);
                 appendedDataView.SetCurrentRowObject(_dst);
+                var appendedIndices = _parent._bindings.AddedColumnIndices;
 
-                Func<int, bool> appendedPredicate =
-                    col =>
-                    {
-                        col = _parent._bindings.AddedColumnIndices[col];
-                        return predicate(col);
-                    };
-
-                _appendedRow = appendedDataView.GetRowCursor(appendedPredicate);
+                var cols = colsNeeded.Select(c => c.Name);
+                _appendedRow = appendedDataView.GetRowCursor(appendedDataView.Schema.Where(c => cols.Contains(c.Name)));
```

Comment:
>cols.Contains(c.Name) [](start = 96, length = 21)

Consider also paying attention only to non-hidden columns, just to make things a bit more restrictive? Does this make sense?

Reply:
Actually we can simplify this is a lot by just having `cols.Select(c => appendedDataView.Schema[c.Name])`. Which is an `IEnumerable` and far more clear about your intentions, I believe.

---
In reply to: [248452752](https://github.com/dotnet/machinelearning/pull/1975#discussion_r248452752) [](ancestors = 248452752)

Reply:
Keeping it to Where, to check for IsHidden. Can't do that with Select. 

---
In reply to: [248453731](https://github.com/dotnet/machinelearning/pull/1975#discussion_r248453731) [](ancestors = 248453731,248452752)

Reply:
Hi @sfilipi , bringiing this up again... Absolutely you can. The string indexer can *never* return a hidden column. That's what hidden means!

Reply:
Keeping it as is for a different reason: the columns need to be selected from the appendedDataview, and not from cols needed, because the indexes are different... 

---
In reply to: [249126662](https://github.com/dotnet/machinelearning/pull/1975#discussion_r249126662) [](ancestors = 249126662)